### ISSUE TITLE
rsync build varies according to kernel cmdline ipv6.disable

### ISSUE THREAD

**Issue by pwaller** [NixOS/nixpkgs#360152]
## Issue description

Rsync is not currently reproducible. Upstream issue: https://github.com/RsyncProject/rsync/issues/675

On nixos-unstable:

```
nix build --rebuild nixpkgs/23e89b7da85c3640bbc2173fe04f4bd114342367#rsync
```

(or `nix-build --check -A rsync` from that commit)

Reports:

```
error: derivation '/nix/store/gppvg32hwnf1h9dvf38mjwfrs74s12jz-rsync-3.3.0.drv' may not be deterministic: output '/nix/store/x850848v3xl4wxjqzc3q9jp7j6fbkh27-rsync-3.3.0' differs from '/nix/store/x850848v3xl4wxjqzc3q9jp7j6fbkh27-rsync-3.3.0.check'
```

Diffoscope output for the two paths show only the binaries differ, and they differ in various virtual addresses: [rsync-diffoscope.txt.gz](https://github.com/user-attachments/files/17961249/rsync-diffoscope.txt.gz)

### Steps to reproduce

* Grab rsync from cache.nixos.org.
* Attempt to rebuild it.

## Technical details

```
 - system: `"x86_64-linux"`
 - host os: `Linux 6.11.9, NixOS, 25.05 (Warbler), 25.05.20241119.23e89b7`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.24.10`
 - nixpkgs: `/nix/store/c9wv7i0af6mysmy65x6nvyfw5izzxv4g-source`
```

**Comment by pwaller** [comment]
I have two machines, one reproduces cache.nixos.org, the other does not.

It looks like cache.nixos.org has `#define INET6 1`, but my local machine doesn't; it looks as though my local machine has `ipv6.disable=1` set but the others do not.

Seems slightly odd to me that the rsync build would check if the local machine has ipv6 enabled.

```
-|  /* AF_INET6 availability check */
-| #include <stdlib.h>
-| #include <sys/types.h>
-| #include <sys/socket.h>
-| int main()
-| {
-|    if (socket(AF_INET6, SOCK_STREAM, 0) < 0)
-|      exit(1);
-|    else
-|      exit(0);
-| }
-|
-configure.sh:7726: result: no
+configure.sh:7718: $? = 0
+configure.sh:7720: result: yes 
```

**Comment by pwaller** [comment]
Upstream issue: https://github.com/RsyncProject/rsync/issues/675

**Comment by pwaller** [comment]
Confirmed that re-enabling ipv6 on the original machine has now caused it to match the cache.nixos.org build.

**Comment by pwaller** [comment]
Per @raboof's suggestion on matrix, `--enable-ipv6` skips the configure check.

**Comment by paparodeo** [comment]
curl has the same check

**Comment by pwaller** [comment]
Hmm. I'm seeing the build id vary in curl so I can't confirm whether the same fix is an improvement there; I'm not seeing the text section vary though, suggesting that at least for now curl is invariant.

**Comment by raboof** [comment]
> Hmm. I'm seeing the build id vary in curl

Interestingly I don't see that here.

**Comment by pwaller** [comment]
Fix for rsync merged in #360445. I haven't time to dig into the curl situation at the moment, if there is one.
